PythonVisualizations/Heap.py

```diff
@@ -0,0 +1,518 @@
+
+try:
+    from drawable import *
+    from VisualizationApp import *
+except ModuleNotFoundError:
+    from .drawable import *
+    from .VisualizationApp import *
+
+
+class Heap(VisualizationApp):
+    nextColor = 0
+    MAX_ARG_WIDTH = 8
+    MAX_SIZE = 31
+    CELL_WIDTH = 25
+    CELL_HEIGHT = 12
+    CELL_BORDER = 1
+    CELL_BORDER_COLOR = 'black'
+    HEAP_X0 = 50
+    HEAP_Y0 = 30
+
+    def __init__(self, size=2, maxArgWidth=MAX_ARG_WIDTH, title="Heap", **kwargs):
+        super().__init__(title=title, maxArgWidth=maxArgWidth, **kwargs)
+        self.size = size
+        self.title = title
+        self.list = []
+        self.maxArgWidth = maxArgWidth
+        self.buttons = self.makeButtons()
+        self.display()
+        self.indexDisplay = self.createIndex(len(self.list))
+
+    def __str__(self):
+        return str(self.list)
+
+    # Create an index arrow to point at an indexed cell with an optional name label
+    def createIndex(self, index, name=None):
+        cell_coords = self.cellCoords(index)
+        cell_center = self.cellCenter(index)
+        x0 = self.HEAP_X0 - self.CELL_WIDTH - 5
+        x1 = self.HEAP_X0 - self.CELL_WIDTH * 3 // 10
+        y0 = y1 = cell_coords[-1] + self.CELL_HEIGHT // 2
+        if not name:
+            label = "nItems"  # labels the top of the Heap "top" with the pointer arrow
+        else:
+            label = name
+        return self.drawArrow(
+            x0, y0, x1, y1, self.VARIABLE_COLOR, self.SMALL_FONT, name=label)
+
+        # draw the actual arrow
+
+    def drawArrow(
+            self, x0, y0, x1, y1, color, font, name=None):
+        arrow = self.canvas.create_line(
+            x0, y0, x1, y1, arrow="last", fill=color)
+        if name:
+            label = self.canvas.create_text(
+                x0 - self.CELL_WIDTH / 2, y0 + (self.CELL_HEIGHT / 16) - 3, text=name, anchor=SW,
+                font=font, fill=color)
+        return (arrow, label) if name else (arrow,)
+
+    def swap(self, a, b, aCellObjects=[], bCellObjects=[]):
+        itemsA = [self.list[a].display_shape,
+                  self.list[a].display_val] + aCellObjects
+        itemsB = [self.list[b].display_shape,
+                  self.list[b].display_val] + bCellObjects
+        upDelta = (0, - self.CELL_HEIGHT * 4 // 3)
+        downDelta = multiply_vector(upDelta, -1)
+        if a == b:  # Swapping with self - just move up & down
+            self.moveItemsBy(itemsA, upDelta, sleepTime=0.02)
+            self.moveItemsBy(itemsA, downDelta, sleepTime=0.02)
+            return
+        # make a and b cells plus their associated items switch places
+        self.moveItemsOnCurve(
+            itemsA + itemsB, [self.canvas.coords(i) for i in itemsB + itemsA],
+            sleepTime=0.05, startAngle=90 * 11 / (10 + abs(a - b)))
+        # perform the actual cell swap operation in the list
+        self.list[a], self.list[b] = self.list[b], self.list[a]
+        # HEAP FUNCTIONALITY

```

Comment:
The `swap()` function is not needed since both sift-up and sift-down operations store one item as they do their work.  See the textbook code.  This saves half the assignment operations.